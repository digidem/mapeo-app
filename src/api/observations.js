// @flow
import fetch from 'fetch';
import { jsonRequest } from './base';

export interface CreateRequest {
  lat: number;
  lon: number;
}

export type UpdateRequest = {
  id: string
};

class Observation {
  static list = () =>
    jsonRequest({
      method: 'GET',
      route: '/observations'
    });

  static create = (observation: CreateRequest) =>
    jsonRequest({
      method: 'POST',
      route: '/observations',
      body: {
        device_id: '1',
        ...observation
      }
    });

  static update = (observation: UpdateRequest) =>
    jsonRequest({
      method: 'PUT',
      route: `/observations/${observation.id}`,
      body: {
        ...observation
      }
    });
}

export default Observation;
