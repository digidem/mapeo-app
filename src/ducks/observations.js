// @flow

import update from 'immutability-helper';
import {keyBy} from 'lodash';
import {create} from '../lib/redux';
import {createSelector} from '../lib/selector';
import type {StoreState} from '../types/redux';
import type {Observation} from '../types/observation';

export const {
  type: OBSERVATION_LIST,
  action: observationList,
  reducer: observationListReducer
} = create('OBSERVATION_LIST', {
  success: (state, action) => {
    const observations = keyBy(action.payload, 'id');

    if (state.observations) {
      Object.keys(state.observations).forEach(id => observations[id] = {
        ...observations[id],
        ...state.observations[id],
      });
    }

    const newState = update(state, {observations: {$set: observations}});

    return newState;
  }
});

export const {
  type: OBSERVATION_CREATE,
  action: observationCreate,
  reducer: observationCreateReducer
} = create('OBSERVATION_CREATE', {
  start: (state, action) => {
    const observation = action.meta;

    if (!observation.lat && !observation.lon && state.gps.data) {
      observation.lat = state.gps.data.latitude;
      observation.lon = state.gps.data.longitude;
    }

    const newState = update(state, {selectedObservation: {$set: observation}});

    return newState;
  },
  success: (state, action) => {
    const newState = update(state, {selectObservation: {$set: action.payload}});

    return newState;
  }
});

export const {
  type: OBSERVATION_UPDATE,
  action: observationUpdate,
  reducer: observationUpdateReducer
} = create('OBSERVATION_UPDATE', {
  start: (state, action) => {
    let newState;
    if (state.selectedObservation &&
        action.meta.id === state.selectedObservation.id) {
      newState = update(state, {
        selectedObservation: {
          $set: {...state.selectedObservation, ...action.meta},
        }
      });
    } else {
      newState = update(state, {
        observations: {
          [action.meta.id]: {
            $set: {...state.observations[action.meta.id], ...action.meta}
          }
        }
      });
    }

    return newState;
  }
});

export const {
  type: OBSERVATION_SELECT,
  action: observationSelect,
  reducer: observationSelectReducer
} = create('OBSERVATION_SELECT', {
  start: (state, action) => {
    if (action.meta === undefined && state.selectedObservation) {
      return update(state, {
        selectedObservation: {$set: action.meta},
        observations: {
          [state.selectedObservation.id]: {
            $set: {
              ...(state.observations[state.selectedObservation.id] || {}),
              ...state.selectedObservation,
            }
          }
        }
      });
    } else {
      return update(state, {selectedObservation: {$set: action.meta}});
    }
  }
});

export const selectObservation = createSelector(
    [(state: StoreState, id: string): Observation =>
         state.app.observations[id]],
    (observation: Observation): Observation => observation,
    (observation: Observation, id: string): string => id);

export default [
  observationListReducer, observationCreateReducer, observationUpdateReducer,
  observationSelectReducer
];
