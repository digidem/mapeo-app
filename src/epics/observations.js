// @flow
import React from 'react';
import { Image } from 'react-native';
import type { ActionsObservable } from 'redux-observable';
import {
  OBSERVATION_LIST,
  observationList,
  OBSERVATION_CREATE,
  observationCreate,
  OBSERVATION_UPDATE,
  observationUpdate
} from '../ducks/observations';
import { Action } from '../types/redux';
import type { Observation as ObservationType } from '../types/observation';
import type { StoreState } from '../types/redux';
import Observation from '../api/observations';
import type { CreateRequest, UpdateRequest } from '../api/observations';

export const observationListEpic = (
  action$: ActionsObservable<Action<string, ObservationType[]>>,
  store: StoreState
) =>
  action$
    .ofType(OBSERVATION_LIST)
    .filter(action => action.status === 'Start')
    .flatMap(() =>
      Observation.list().map(observations => {
        console.log('RN - Observations List', observations);
        return observationList('', observations);
      })
    );

export const observationCreateEpic = (
  action$: ActionsObservable<Action<CreateRequest, ObservationType>>,
  store: StoreState
) =>
  action$
    .ofType(OBSERVATION_CREATE)
    .filter(action => action.status === 'Start')
    .flatMap(action =>
      Observation.create(action.meta).map(observation => {
        console.log('RN - Observation Created');
        return observationCreate(action.meta, observation);
      })
    );

// export const observationUpdateEpic = (
//   action$: ActionsObservable<Action<UpdateRequest, ObservationType>>,
//   store: StoreState
// ) =>
//   action$
//     .ofType(OBSERVATION_UPDATE)
//     .filter(action => action.status === 'Start')
//     .flatMap(action =>
//       Observation.update(action.meta).map(observation => {
//         console.log('RN - Observation Updated - ', action.meta, observation);
//         return observationUpdate(action.meta, observation);
//       })
//     );

export default [
  observationListEpic,
  observationCreateEpic
  // observationUpdateEpic
];
