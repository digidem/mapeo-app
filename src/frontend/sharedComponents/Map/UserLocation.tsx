import MapboxGL from "@react-native-mapbox-gl/maps";
import * as React from "react";
import { useExperiments } from "../../hooks/useExperiments";
interface UserLocationProps {
  visible: boolean;
  minDisplacement: number;
}

export const UserLocation = ({
  visible,
  minDisplacement,
}: UserLocationProps) => {
  const [{ directionalArrow }] = useExperiments();

  return (
    <MapboxGL.UserLocation
      visible={visible}
      minDisplacement={directionalArrow ? 1 : minDisplacement}
      showsUserHeadingIndicator={directionalArrow}
    />
  );
};
