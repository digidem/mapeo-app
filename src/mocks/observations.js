// @flow

export const createObservation = (observation?: any) => ({
  type: 'Observation',
  id: 1,
  lat: 50.5,
  lon: 50.5,
  link: 'linkID',
  created: new Date('2018-02-25T20:23:00.000Z'),
  name: 'Oil Spill',
  notes:
    'This is pretty bad. It smells pretty noxious, and the entire pond is affected.',
  observedBy: 'You',
  media: [],
  icon: require('../images/categories/category_18.png'),
  categoryId: 19,

  ...observation
});
