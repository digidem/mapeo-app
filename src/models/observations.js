// @flow
import type { Observation } from '../types/observation';

export const applyObservationDefaults = (partial: Object): Observation => ({
  type: 'Rios y corrientes',
  lat: 0,
  lon: 0,
  link: 'link',
  created: new Date(),
  name: '',
  notes: '',
  observedBy: 'TÃº',
  media: [],
  icon: null,
  categoryId: '0',
  fields: [],
  mediaBackup: [],

  ...partial
});
