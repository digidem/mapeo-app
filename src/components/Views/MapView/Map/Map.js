// @flow
import React from 'react';
import {
  Image,
  ImageBackground,
  StyleSheet,
  View,
  TouchableOpacity,
  Dimensions,
  PermissionsAndroid
} from 'react-native';
import { NavigationActions, withNavigationFocus } from 'react-navigation';
import MapboxGL from '@mapbox/react-native-mapbox-gl';
import Icon from 'react-native-vector-icons/MaterialIcons';
import type { Observation } from '@types/observation';
import { isEmpty, size, map, filter } from 'lodash';
import env from '../../../../../env.json';

import AddButton from '../../../../images/add-button.png';
import Gradient from '../../../../images/gradient-overlay.png';
import { WHITE } from '../../../../lib/styles';
import { applyObservationDefaults } from '../../../../models/observations';
import type { Resource } from '../../../../types/redux';
import type { GPSState } from '../../../../types/gps';
import type { Style } from '../../../../types/map';

export type StateProps = {
  observations: {
    [id: string]: Observation
  },
  selectedObservation?: Observation,
  gps?: GPSState,
  selectedStyle?: Style
};

export type DispatchProps = {
  listObservations: () => void,
  createObservation: (observation: Observation) => void,
  updateObservation: (observation: Observation) => void,
  selectObservation: (observation: Observation) => void,
  updateObservationSource: () => void,
  listStyles: () => void
};

type Props = {
  navigation: NavigationActions
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'flex-start',
    alignItems: 'center',
    backgroundColor: 'white'
  },

  mapPlaceholder: {
    backgroundColor: '#000',
    color: '#FFF',
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center'
  },

  map: {
    flex: 1,
    alignSelf: 'stretch'
  },

  newObservation: {
    alignSelf: 'center'
  },

  buttons: {
    flexDirection: 'row'
  },

  btn: {
    backgroundColor: '#4444FF',
    padding: 10,
    margin: 10
  },

  btnText: {
    color: 'white'
  },

  info: {
    textAlign: 'left',
    color: '#333333',
    marginBottom: 5
  }
});

const mapboxStyles = MapboxGL.StyleSheet.create({
  point: {
    circleColor: '#5086EC',
    circleRadius: 5,
    circleStrokeColor: '#fff',
    circleStrokeWidth: 2
  },
  observation: {
    circleColor: '#F29D4B',
    circleRadius: 5,
    circleStrokeColor: '#fff',
    circleStrokeWidth: 2
  },
  currentLocation: {
    circleColor: '#2657B1',
    circleRadius: 7,
    circleStrokeColor: '#fff',
    circleStrokeWidth: 3
  }
});

class Map extends React.Component<Props & StateProps & DispatchProps> {
  constructor() {
    super();

    MapboxGL.setAccessToken(env.accessToken);
  }

  componentDidMount() {
    const {
      observations,
      listObservations,
      selectedStyle,
      listStyles
    } = this.props;

    listObservations();

    if (!selectedStyle) {
      listStyles();
    }
  }

  shouldComponentUpdate(nextProps: Props & StateProps & DispatchProps) {
    if (nextProps.navigation.isFocused()) {
      nextProps.listObservations();
      return nextProps !== this.props;
    }

    return false;
  }

  map: any;

  handleCreateObservation = () => {
    const {
      createObservation,
      observations,
      updateObservation,
      selectedObservation,
      updateObservationSource,
      selectedStyle,
      navigation
    } = this.props;
    const initialObservation = applyObservationDefaults({
      id: size(observations) + 1
    });

    navigation.navigate({
      routeName: 'Categories',
      key: 'CategoriesView'
    });
    updateObservationSource();
    createObservation(initialObservation);
  };

  handleObservationPress = (id: string) => {
    const { observations, selectObservation, navigation } = this.props;

    if (observations[id]) {
      selectObservation(observations[id]);
      navigation.navigate({ routeName: 'ObservationDetailView' });
    }
  };

  handleLongPress = (point: Object) => {
    const { coordinates } = point.geometry;
    const {
      createObservation,
      navigation,
      observations,
      updateObservationSource
    } = this.props;
    const initialObservation = applyObservationDefaults({
      id: size(observations) + 1,
      lat: coordinates[1],
      lon: coordinates[0]
    });

    navigation.navigate({
      routeName: 'Categories',
      key: 'CategoriesView'
    });
    updateObservationSource();
    createObservation(initialObservation);
  };

  handleMapViewRef = c => {
    this.map = c;
  };

  goToCurrentLocation = () => {
    const { gps } = this.props;

    if (this.map && gps) {
      this.map.moveTo([gps.longitude, gps.latitude], 200);
    }
  };

  render() {
    const { observations, gps, selectedStyle } = this.props;

    return (
      <View
        style={{
          width: Dimensions.get('window').width,
          height: Dimensions.get('window').height
        }}
      >
        <View style={{ flex: 1 }}>
          <MapboxGL.MapView
            style={{ flex: 1 }}
            centerCoordinate={gps ? [gps.longitude, gps.latitude] : undefined}
            ref={this.handleMapViewRef}
            zoomLevel={
              selectedStyle
                ? Math.floor(
                    (selectedStyle.minzoom + selectedStyle.maxzoom) / 2
                  )
                : 12
            }
            minZoomLevel={selectedStyle ? selectedStyle.minzoom : undefined}
            maxZoomLevel={selectedStyle ? selectedStyle.maxzoom : undefined}
            logoEnabled
            pitchEnabled={false}
            rotateEnabled={false}
            onLongPress={this.handleLongPress}
            compassEnabled={false}
            styleURL={
              selectedStyle
                ? `http://localhost:9080/styles/${selectedStyle.id}/style.json`
                : undefined
            }
          >
            {!!observations && !isEmpty(observations)
              ? map(observations, (o: Observation) => (
                <MapboxGL.ShapeSource
                  onPress={() => this.handleObservationPress(o.id)}
                  key={o.id}
                  id={`observations-${o.id}`}
                  shape={{
                      type: 'Feature',
                      geometry: {
                        type: 'Point',
                        coordinates: [o.lon, o.lat]
                      },
                      properties: {
                        name: o.name
                      }
                    }}
                >
                  <MapboxGL.CircleLayer
                    id={`circles-${o.id}`}
                    style={mapboxStyles.observation}
                  />
                </MapboxGL.ShapeSource>
                ))
              : null}
            {gps && (
              <MapboxGL.ShapeSource
                key="current-location"
                id="current-location"
                shape={{
                  type: 'Feature',
                  geometry: {
                    type: 'Point',
                    coordinates: [gps.longitude, gps.latitude]
                  }
                }}
              >
                <MapboxGL.CircleLayer
                  id="circles-current-location"
                  style={mapboxStyles.currentLocation}
                />
              </MapboxGL.ShapeSource>
            )}
          </MapboxGL.MapView>
          <View
            style={{
              position: 'absolute',
              bottom: 25,
              alignItems: 'center',
              backgroundColor: 'transparent',
              zIndex: 5,
              flexDirection: 'row',
              justifyContent: 'space-around',
              width: Dimensions.get('window').width
            }}
          >
            <View style={{ width: 50 }} />
            <TouchableOpacity
              onPress={this.handleCreateObservation}
              style={styles.newObservation}
            >
              <Image
                source={AddButton}
                style={{
                  width: 125,
                  height: 125
                }}
              />
            </TouchableOpacity>
            <TouchableOpacity onPress={this.goToCurrentLocation}>
              <Icon color={WHITE} name="my-location" size={50} />
            </TouchableOpacity>
          </View>
        </View>
        <View
          style={{
            position: 'absolute',
            left: 0,
            backgroundColor: 'transparent'
          }}
        >
          <ImageBackground
            source={Gradient}
            style={{
              width: Dimensions.get('window').width,
              height: 100
            }}
          />
        </View>
      </View>
    );
  }
}

export default withNavigationFocus(Map);
