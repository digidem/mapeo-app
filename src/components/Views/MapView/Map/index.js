// @flow
import { connect } from 'react-redux';
import type { Dispatch } from 'redux';
import {
  observationCreate,
  observationUpdate,
  observationSelect
} from '../../../../ducks/observations';
import { observationSource } from '../../../../ducks/observationSource';
import { StoreState } from '../../../../types/redux';
import Map from './Map';
import type { DispatchProps, StateProps } from './Map';

const mapStateToProps = (state: StoreState) => ({
  observations: state.observations,
  selectedObservation: state.selectedObservation,
  coords: state.gps.coords,
  selectedStyle: state.map.selectedStyle
});

function mapDispatchToProps(dispatch: Dispatch<*>) {
  return {
    createObservation: observation => dispatch(observationCreate(observation)),
    updateObservation: observation => dispatch(observationUpdate(observation)),
    selectObservation: o => dispatch(observationSelect(o)),
    updateObservationSource: () => dispatch(observationSource('map'))
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Map);
