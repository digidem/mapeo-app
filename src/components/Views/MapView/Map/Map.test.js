// @flow
import React from 'react';
import 'react-native';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Map from './Map';
import { createObservation } from '../../../../mocks/observations';

describe('MapView tests', () => {
  test('snapshots', () => {
    const observation = createObservation();
    const isFocused = () => true;
    const addListener = () => true;
    const listObservations = jest.fn();
    const createObservationMock = jest.fn();
    const updateObservation = jest.fn();
    const selectObservation = jest.fn();
    const updateObservationSource = jest.fn();
    const listStyles = jest.fn();

    let tree;
    tree = shallow(
      <Map
        observations={{}}
        navigation={{ isFocused, addListener }}
        selectedObservation={observation}
        listObservations={listObservations}
        createObservation={createObservationMock}
        updateObservation={updateObservation}
        selectObservation={selectObservation}
        updateObservationSource={updateObservationSource}
        listStyles={listStyles}
      />
    );
    expect(tree).toMatchSnapshot();
  });
});
