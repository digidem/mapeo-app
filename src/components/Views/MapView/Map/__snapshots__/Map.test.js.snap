// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapView tests snapshots 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <withNavigation(withNavigationFocus(Map))
    createObservation={[MockFunction]}
    listObservations={[MockFunction]}
    listStyles={[MockFunction]}
    navigation={
      Object {
        "addListener": [Function],
        "isFocused": [Function],
      }
    }
    observations={Object {}}
    selectObservation={[MockFunction]}
    selectedObservation={
      Object {
        "attachments": Array [],
        "categoryId": 19,
        "created": 2018-02-25T20:23:00.000Z,
        "fields": Array [
          Object {
            "answer": "",
            "answered": false,
            "id": "0",
            "name": "fieldName",
            "placeholder": "this is some placeholder text",
            "type": "text",
          },
        ],
        "icon": Object {
          "testUri": "../../../src/images/categories/category_18.png",
        },
        "id": 1,
        "lat": 50.5,
        "link": "linkID",
        "lon": 50.5,
        "name": "Oil Spill",
        "notes": "This is pretty bad. It smells pretty noxious, and the entire pond is affected.",
        "observedBy": "You",
        "type": "Observation",
      }
    }
    updateObservation={[MockFunction]}
    updateObservationSource={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": [Function],
    },
    "ref": null,
    "rendered": [Function],
    "type": Object {
      "$$typeof": Symbol(react.context),
      "Consumer": [Circular],
      "Provider": Object {
        "$$typeof": Symbol(react.provider),
        "_context": [Circular],
      },
      "_calculateChangedBits": null,
      "_changedBits": 0,
      "_changedBits2": 0,
      "_currentRenderer": null,
      "_currentRenderer2": null,
      "_currentValue": undefined,
      "_currentValue2": undefined,
      "_defaultValue": undefined,
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": [Function],
      },
      "ref": null,
      "rendered": [Function],
      "type": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": [Circular],
        "Provider": Object {
          "$$typeof": Symbol(react.provider),
          "_context": [Circular],
        },
        "_calculateChangedBits": null,
        "_changedBits": 0,
        "_changedBits2": 0,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": undefined,
        "_currentValue2": undefined,
        "_defaultValue": undefined,
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
